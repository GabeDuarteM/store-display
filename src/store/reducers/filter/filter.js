import {
  FILTER__CHANGE_GENDER,
  FILTER__CHANGE_SORT,
  FILTER__LOAD_MORE_PRODUCTS,
} from './filterActions'

const returnInitialState = () => ({
  gender: 'woman',
  sort: 'price-asc',
  productListLength: 9,
})

const filter = (state = returnInitialState(), action = {}) => {
  switch (action.type) {
    case FILTER__CHANGE_GENDER:
      return {
        ...state,
        gender: action.payload,
        productListLength: 9,
      }
    case FILTER__CHANGE_SORT:
      return {
        ...state,
        sort: action.payload,
      }
    case FILTER__LOAD_MORE_PRODUCTS:
      return {
        ...state,
        productListLength: state.productListLength + 9,
      }
    default:
      return state
  }
}

export default filter
