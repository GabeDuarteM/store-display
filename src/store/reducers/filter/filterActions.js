export const FILTER__CHANGE_GENDER = 'FILTER/CHANGE_GENDER'
export const FILTER__CHANGE_SORT = 'FILTER/CHANGE_SORT'
export const FILTER__LOAD_MORE_PRODUCTS = 'FILTER/LOAD_MORE_PRODUCTS'

export const filterChangeGender = gender => ({
  type: FILTER__CHANGE_GENDER,
  payload: gender,
})

export const filterChangeSort = sort => ({
  type: FILTER__CHANGE_SORT,
  payload: sort,
})

export const filterLoadMoreProducts = () => ({
  type: FILTER__LOAD_MORE_PRODUCTS,
})
