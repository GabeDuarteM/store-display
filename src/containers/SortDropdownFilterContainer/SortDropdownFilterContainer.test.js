import React from 'react'

import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import SortDropdownFilterContainerConnected, {
  SortDropdownFilterContainer,
  propTypes,
} from './SortDropdownFilterContainer'
import { filterChangeSort } from '../../store/reducers/filter/filterActions'

const createWrapper = func =>
  func(
    <SortDropdownFilterContainer
      selectedItem={{ id: 'price-asc', text: 'Price: low to high' }}
      onChange={() => {}}
    />,
  )

const createWrapperShallow = () => createWrapper(shallow)

const createWrapperMount = () => createWrapper(mount)

describe('COMPONENT: <SortDropdownFilterContainer />', () => {
  it('should render without crashing', () => {
    createWrapperMount()
  })

  it("should match component's snapshot", () => {
    const wrapper = createWrapperShallow()

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it("should match component's propTypes snapshot", () => {
    expect(propTypes).toMatchSnapshot()
  })

  it('should start with the state menuIsOpen = false', () => {
    const wrapper = createWrapperShallow()
    const { menuIsOpen } = wrapper.state()

    expect(menuIsOpen).toBeFalsy()
  })

  it('should change the menuIsOpen state when handleChangeMenuIsOpen is called', () => {
    const expected = true
    const wrapper = createWrapperShallow()
    wrapper.instance().handleChangeMenuIsOpen(expected)
    const actual = wrapper.state('menuIsOpen')

    expect(actual).toEqual(expected)
  })

  it('should have the correct set of items', () => {
    const expected = [
      { id: 'price-asc', text: 'Price: low to high' },
      { id: 'price-desc', text: 'Price: high to low' },
      { id: 'rating-asc', text: 'Rating: low to high' },
      { id: 'rating-desc', text: 'Rating: high to low' },
    ]
    const wrapper = createWrapperShallow()
    const received = wrapper.prop('items')

    expect(expected).toEqual(received)
  })

  it('should have the correct label', () => {
    const expected = 'Sort by'
    const wrapper = createWrapperShallow()
    const received = wrapper.prop('label')

    expect(expected).toEqual(received)
  })

  it('should have a connected component rendering the props correctly', () => {
    const filter = {
      gender: 'man',
      productListLength: 9,
      sort: 'price-desc',
    }
    const selectedItem = { id: 'price-desc', text: 'Price: high to low' }
    const store = {
      getState: () => ({
        filter,
      }),
      dispatch: jest.fn(),
      subscribe: () => {},
    }

    const component = shallow(
      <SortDropdownFilterContainerConnected store={store} />,
    ).find(SortDropdownFilterContainer)
    const selectedItemProp = component.prop('selectedItem')
    const onChangeProp = component.prop('onChange')
    onChangeProp(selectedItem)

    expect(selectedItemProp).toEqual(selectedItem)
    expect(store.dispatch).toHaveBeenCalled()
    expect(store.dispatch).toHaveBeenCalledWith(
      filterChangeSort(selectedItem.id),
    )
  })

  it('should have a connected component that scrolls to the top when onChange is fired', () => {
    const filter = {
      gender: 'man',
      productListLength: 9,
      sort: 'price-desc',
    }
    const selectedItem = { id: 'price-desc', text: 'Price: high to low' }
    const store = {
      getState: () => ({
        filter,
      }),
      dispatch: jest.fn(),
      subscribe: () => {},
    }

    const component = shallow(
      <SortDropdownFilterContainerConnected store={store} />,
    ).find(SortDropdownFilterContainer)
    const onChangeProp = component.prop('onChange')
    window.scrollTo.mockClear()

    onChangeProp(selectedItem)

    expect(window.scrollTo).toHaveBeenCalledTimes(1)
    expect(window.scrollTo).toHaveBeenLastCalledWith(0, 0)
  })
})
