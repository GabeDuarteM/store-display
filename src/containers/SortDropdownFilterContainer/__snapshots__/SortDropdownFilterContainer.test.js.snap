// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT: <SortDropdownFilterContainer /> should match component's propTypes snapshot 1`] = `
Object {
  "onChange": [Function],
  "selectedItem": [Function],
}
`;

exports[`COMPONENT: <SortDropdownFilterContainer /> should match component's snapshot 1`] = `
<DropdownFilter
  handleChangeMenuIsOpen={[Function]}
  items={
    Array [
      Object {
        "id": "price-asc",
        "text": "Price: low to high",
      },
      Object {
        "id": "price-desc",
        "text": "Price: high to low",
      },
      Object {
        "id": "rating-asc",
        "text": "Rating: low to high",
      },
      Object {
        "id": "rating-desc",
        "text": "Rating: high to low",
      },
    ]
  }
  label="Sort by"
  menuIsOpen={false}
  onChange={[Function]}
  selectedItem={
    Object {
      "id": "price-asc",
      "text": "Price: low to high",
    }
  }
  width={200}
/>
`;
